### PR TITLE
Prototyping Rebalance Protocol

### PR DESCRIPTION
**Pull request by philipnee**
*More detailed description of your change,
if necessary. The PR title and PR message become
the squashed commit message, so use a separate
comment to ping reviewers.*

*Summary of testing strategy (including rationale)
for the feature or bug fix. Unit and/or integration
tests are expected for any behaviour change and
system tests should be considered for larger changes.*

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/ProtocolRequestManager.java [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+        eventQueue.add(new RebalanceCallbackEvent());
+    }
+
+    class RebalanceCallbackEvent extends ApplicationEvent {
+        protected RebalanceCallbackEvent() {
+            super(null); // it should actually accept a type.REBALANCE_CALLBACK_TRIGGER but i'm too lazy to create one
+        }
+    }
+
+    class HeartbeatRequestManager implements RequestManager {
```
Comment (philipnee) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/ProtocolRequestManager.java:53]: obviously not beloing here, but doing this as a prototype

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/ProtocolRequestManager.java [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+            // just a demo, don't do anything
+            return null;
+        }
+
+        public void sendHeartbeat(long currentTimeMs) {
+            heartbeatRequestState.canSendRequest(currentTimeMs);
+        }
+    }
+
+    class ServierSideProtocol implements RebalanceProtocol {
```
Comment (philipnee) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/ProtocolRequestManager.java]: same, just a prototype: as you can see, nothing was implemented

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/GroupState.java [hunk 3]
```diff
@@ -27,6 +27,15 @@ public class GroupState {
     public final String groupId;
     public final Optional<String> groupInstanceId;
     public Generation generation = Generation.NO_GENERATION;
+    public State state = State.UNJOINED;
+
+    enum State {
+        UNJOINED,
+        PREPARE,
+        ASSIGNING,
+        COMPLETE,
```
Comment (guozhangwang) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/GroupState.java:36]: What's the difference between COMPLETE and STABLE?

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/ProtocolRequestManager.java [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+                return protocol.onAssign(currentTimeMs);
+            case COMPLETE:
+                return protocol.onComplete(currentTimeMs);
+            case STABLE:
+                return protocol.onStable(currentTimeMs);
+        }
+        return null;
+    }
+
+    public void ackCallbackInvocation(long currentTimeMs) {
```
Comment (guozhangwang) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/ProtocolRequestManager.java:33]: Not clear when this func would be called?

**Pull request closed without merging**